src/Core/Services/MetadataProviders/CosmosSqlMetadataProvider.cs

```diff
@@ -55,12 +55,13 @@ public class CosmosSqlMetadataProvider : ISqlMetadataProvider
 
         public CosmosSqlMetadataProvider(RuntimeConfigProvider runtimeConfigProvider, IFileSystem fileSystem)
         {
-            _fileSystem = fileSystem;
-            _runtimeConfig = runtimeConfigProvider.GetConfig();
+            RuntimeConfig runtimeConfig = runtimeConfigProvider.GetConfig();
 
-            _databaseType = _runtimeConfig.DataSource.DatabaseType;
+            _fileSystem = fileSystem;
+            _runtimeConfigProvider = runtimeConfigProvider;
+            _databaseType = runtimeConfig.DataSource.DatabaseType;
 
-            CosmosDbNoSQLDataSourceOptions? cosmosDb = _runtimeConfig.DataSource.GetTypedOptions<CosmosDbNoSQLDataSourceOptions>();
+            CosmosDbNoSQLDataSourceOptions? cosmosDb = runtimeConfig.DataSource.GetTypedOptions<CosmosDbNoSQLDataSourceOptions>();
```

Comment:
I'm not really following what the benefit is of changing this constructor's code from:
```csharp
_runtimeConfig = runtimeConfigProvider.GetConfig();
ParseSchemaGraphQLFieldsForJoins();
```
to
```csharp
RuntimeConfig runtimeConfig = runtimeConfigProvider.GetConfig();
_runtimeConfigProvider = runtimeConfigProvider;
ParseSchemaGraphQLFieldsForJoins(runtimeConfig);
```

This constructor is executed a single time during startup, so if you hot reload the config, wouldn't you need to create a new instance of `CosmosSqlMetadataProvider`? I don't see other references to code which calls `ParseSchemaGraphQLFieldsForJoins(...)`

This class is a good use case for understanding how your hot reload event notification system is going to work. When hot-reload occurs, this class just needs to update its stateful `_runtimeConfig` and run any helper functions that setup the remaining state of the metadata provider.

With that in mind, you wouldn't need to make the other changes you've added to this file. e.g. instead of pulling a new runtimeconfig object for each of this class's public functions, each function would, by-design, be using the stateful `_runtimeConfig` member to determine what data to return, which is guaranteed to be the latest considering an implemented mechanism to handle hot reload events.

Reply:
We spoke offline but just to add some context here; the idea was to be defensive for the possible case where we have a hot-reload happen without the cosmosMetaDataProvider being updated or replaced. However, since the fields that store states which are based on the runtimeconfig are all instantiated and set during construction, we should keep that state consistent through the life of the cosmosMetaDataProvider until a hot-reload happens. When the hot-reload happens we then completely update those states to match the new runtimeConfig, likely by creating an entirely new cosmosMetaDataProvider.

Now, to ensure that we have consistency, we will make a private, read only copy of the fields that we use from the runtimeconfig during construction. We also save the runtimeConfigProvider so that we can get up to date information for the public calls that use information from the runtimeConfig that does not set the state of the cosmosMetaDataProvider at time of construction, eg isDevelopmentMode()